pkg/services/serviceaccounts/extsvcaccounts/service.go

```diff
@@ -380,6 +380,56 @@ func (esa *ExtSvcAccountsService) getExtSvcAccountToken(ctx context.Context, org
 	return newKeyInfo.ClientSecret, nil
 }
 
+func genTokenWithRetries(ctxLogger log.Logger, extSvcSlug string) (satokengen.KeyGenResult, error) {
+	var newKeyInfo satokengen.KeyGenResult
+	var err error
+	var ok bool
+	retry := 0
+	for !ok && retry < maxTokenGenRetries {
+		newKeyInfo, err = satokengen.New(extSvcSlug)
+		if err != nil {
+			return satokengen.KeyGenResult{}, err
+		}
+
+		if !strings.Contains(newKeyInfo.ClientSecret, "\x00") {
+			ok = true
+			continue
+		}
+
+		retry++
+
+		ctxLogger.Warn("Generated a token containing NUL, retrying",
+			"service", extSvcSlug,
+			"retry", retry,
+		)
+		// On first retry, log the token parts that contain a nil byte
+		if retry == 1 {
+			logTokenNULParts(ctxLogger, extSvcSlug, newKeyInfo.ClientSecret)
+		}
+	}
+
+	if !ok {
+		return satokengen.KeyGenResult{}, ErrCredentialsGenFailed.Errorf("Failed to generate a token for %s", extSvcSlug)
+	}
+	return newKeyInfo, nil
+}
+
+// logTokenNULParts logs a warning if the external service token contains a nil byte
+// Tokens normally have 3 parts "gl+serviceID_secret_checksum"
+// Log the part of the generated token that contains a nil byte
+func logTokenNULParts(ctxLogger log.Logger, extSvcSlug string, token string) {
+	parts := strings.Split(token, "_")
+	for i := range parts {
+		if strings.Contains(parts[i], "\x00") {
+			ctxLogger.Warn("Token contains NUL",
+				"service", extSvcSlug,
+				"part", i+1,
+				"parts_count", len(parts),
```

Comment:
Do you think this could be useful too?
```suggestion
				"part_len", len(parts[i]),
				"parts_count", len(parts),
```